### PR TITLE
Modify rule S3590: LaYC format 

### PR DESCRIPTION
**Pull request by LoghinVladDev**
## Review

A dedicated reviewer checked the rule description successfully for:

- [ ] logical errors and incorrect information
- [ ] information gaps and missing content
- [ ] text style and tone
- [ ] PR summary and labels follow [the guidelines](https://github.com/SonarSource/rspec/#to-modify-an-existing-rule)

### CODE REVIEW

File: rules/S3590/cfamily/rule.adoc [hunk 1]
```diff
@@ -1,34 +1,137 @@
...
 == Why is this an issue?
 
-Stack allocated memory, like memory allocated with the functions ``++alloca++``, ``++_alloca++``, ``++_malloca++``, ``++__builtin_alloca++``, is automatically released at the end of the function, and should not be released with ``++free++``. Explicitly ``++free++``-ing such memory results in undefined behavior.
+The `free` function and `delete` operator are used exclusively to release dynamically allocated memory.
+Attempting to release any other type of memory is _undefined behavior_.
+
+This rule raises an issue when trying to release pointers to non-heap memory, such as:
+
+* Stack allocated memory - local variables or memory allocated with the `alloca`, `_alloca`, `_malloca` and `__builtin_alloca` functions.
+* Program executable code - function pointers.
```
Comment (pdschbrt) [inline rules/S3590/cfamily/rule.adoc:11]:
```suggestion
* Executable program code - function pointers.
```

File: rules/S3590/cfamily/rule.adoc [hunk 2]
```diff
@@ -1,34 +1,137 @@
...
+* Program data - global and static variables.
+* Read-only program data - constants and strings.
+
```
Comment (pdschbrt) [inline rules/S3590/cfamily/rule.adoc:14]: We also need the `What is the potential impact` section in between.

File: rules/S3590/cfamily/rule.adoc [hunk 3]
```diff
@@ -1,34 +1,137 @@
+Explicitly releasing non-heap memory leads to undefined behavior.
+
 == Why is this an issue?
 
-Stack allocated memory, like memory allocated with the functions ``++alloca++``, ``++_alloca++``, ``++_malloca++``, ``++__builtin_alloca++``, is automatically released at the end of the function, and should not be released with ``++free++``. Explicitly ``++free++``-ing such memory results in undefined behavior.
+The `free` function and `delete` operator are used exclusively to release dynamically allocated memory.
+Attempting to release any other type of memory is _undefined behavior_.
+
+This rule raises an issue when trying to release pointers to non-heap memory, such as:
```
Comment (pdschbrt) [inline rules/S3590/cfamily/rule.adoc:8]: Please try to rephrase this a bit. I wouldn't explicitly mention that the "rule raises" since the rule has already been raised if a developer reads this. (The following list is really nice, it's just about the phrasing.)

File: rules/S3590/cfamily/rule.adoc [hunk 4]
```diff
@@ -1,34 +1,137 @@
...
+
+* Stack allocated memory - local variables or memory allocated with the `alloca`, `_alloca`, `_malloca` and `__builtin_alloca` functions.
+* Program executable code - function pointers.
+* Program data - global and static variables.
+* Read-only program data - constants and strings.
+
+
+== How to fix it
+
+Remove any calls to `free` or `delete` that release non-heap memory.
```
Comment (pdschbrt) [inline rules/S3590/cfamily/rule.adoc:18]:
```suggestion
Remove any calls to `free` or `delete` that aim at releasing non-heap memory.
```

File: rules/S3590/cfamily/rule.adoc [hunk 5]
```diff
@@ -1,34 +1,137 @@
...
+
+As a best practice, do not use the same pointer variable to point to heap memory and any other type of memory at different points in the program.
```
Comment (pdschbrt) [inline rules/S3590/cfamily/rule.adoc:20]: Please provide a more elaborate description of what you mean here and also give a code example, and then put it in the `Going the extra mile` subsection that follows the `Code examples` subsection.

File: rules/S3590/cfamily/rule.adoc [hunk 6]
```diff
@@ -1,34 +1,137 @@
...
+
+==== Noncompliant code example
+
+[source,cpp,diff-type=noncompliant,diff-id=1]
+----
+void fun(int size) {
+  int number = 0;
+  char* name = (char*)alloca(size);
+  // ...
+  delete &number; // Noncompliant, memory allocated on the stack.
```
Comment (pdschbrt) [inline rules/S3590/cfamily/rule.adoc:34]:
```suggestion
  delete &number; // Noncompliant: memory is stack-allocated.
```

File: rules/S3590/cfamily/rule.adoc [hunk 7]
```diff
@@ -1,34 +1,137 @@
...
+  free(name); // Noncompliant, memory allocated on the stack.
```
Comment (pdschbrt) [inline rules/S3590/cfamily/rule.adoc:35]:
```suggestion
  free(name); // Noncompliant: memory is stack-allocated.
```

File: rules/S3590/cfamily/rule.adoc [hunk 8]
```diff
@@ -1,34 +1,137 @@
...
+----
+
+==== Compliant solution
+
+[source,cpp,diff-type=compliant,diff-id=1]
+----
+void fun(int size) {
+  int number = 0;
+  char* name = (char*)alloca(size);
+  // ...
```
Comment (pdschbrt) [inline rules/S3590/cfamily/rule.adoc:46]: At a comment at the end of this function like `Compliant: stack memory is automatically released once the variables go out of scope` or something along those lines.

File: rules/S3590/cfamily/rule.adoc [hunk 9]
```diff
@@ -1,34 +1,137 @@
...
+}
+----
+
+Program executable code:
+
+==== Noncompliant code example
+
+[source,cpp,diff-type=noncompliant,diff-id=2]
+----
+int get10() {
```
Comment (pdschbrt) [inline rules/S3590/cfamily/rule.adoc:56]:
```suggestion
int getValue() {
```

File: rules/S3590/cfamily/rule.adoc [hunk 10]
```diff
@@ -1,34 +1,137 @@
...
+  return 10;
+}
+
+void fun() {
+  // ...
+  free((void*) &get10); // Noncompliant, memory part of program executable code.
```
Comment (pdschbrt) [inline rules/S3590/cfamily/rule.adoc:62]:
```suggestion
  free((void*) &getValue); // Noncompliant: memory is part of executable code.
```

File: rules/S3590/cfamily/rule.adoc [hunk 11]
```diff
@@ -1,34 +1,137 @@
...
+}
+----
 
+==== Compliant solution
+
+[source,cpp,diff-type=compliant,diff-id=2]
+----
+int get10() {
```
Comment (pdschbrt) [inline rules/S3590/cfamily/rule.adoc:70]: Adjust accordingly and also at some sort of `Compliant:` comment.

File: rules/S3590/cfamily/rule.adoc [hunk 12]
```diff
@@ -1,34 +1,137 @@
...
+  static inline int data = 64;
+};
+
 void fun() {
-  char *name = (char *) alloca(size);
+  static int x = 8;
   // ...
-  free(name); // Noncompliant, memory allocated on the stack
-  char *name2 = "name";
+  free(&x); // Noncompliant, memory part of program data.
```
Comment (pdschbrt) [inline rules/S3590/cfamily/rule.adoc:92]:
```suggestion
  free(&x); // Noncompliant: memory part of the program's data.
```

File: rules/S3590/cfamily/rule.adoc [hunk 13]
```diff
@@ -1,34 +1,137 @@
...
+  free(&S::data); // Noncompliant, memory part of program data.
```
Comment (pdschbrt) [inline rules/S3590/cfamily/rule.adoc:93]:
```suggestion
  free(&S::data); // Noncompliant: memory part of the program's data.
```

File: rules/S3590/cfamily/rule.adoc [hunk 14]
```diff
@@ -1,34 +1,137 @@
...
+
+==== Compliant solution
+
+[source,cpp,diff-type=compliant,diff-id=3]
+----
+struct S {
+  static inline int data = 64;
+};
+
+void fun() {
```
Comment (pdschbrt) [inline rules/S3590/cfamily/rule.adoc:105]: Add a `Compliant:` comment and state that global data does not need to be manually freed.

File: rules/S3590/cfamily/rule.adoc [hunk 15]
```diff
@@ -1,34 +1,137 @@
...
+[source,cpp,diff-type=noncompliant,diff-id=4]
 ----
+int const limit = 128;
+
 void fun() {
-  char *name = (char *) alloca(size);
+  char* name = "string";
   // ...
-  char *name2 = "name";
+  free((void*)&limit); // Noncompliant, memory part of read-only program data.
```
Comment (pdschbrt) [inline rules/S3590/cfamily/rule.adoc:122]:
```suggestion
  free((void*)&limit); // Noncompliant: memory part of program's read-only data.
```

File: rules/S3590/cfamily/rule.adoc [hunk 16]
```diff
@@ -1,34 +1,137 @@
...
+  free(name); // Noncompliant, memory part of read-only program data.
```
Comment (pdschbrt) [inline rules/S3590/cfamily/rule.adoc:123]:
```suggestion
  free(name); // Noncompliant: memory part of read-only program data.
```

Comment (pdschbrt) [inline rules/S3590/cfamily/rule.adoc:123]: `Name` is a stack variable here.

**Review by pdschbrt — CHANGES_REQUESTED** [review]
Found a few minor things, but nothing major.

File: rules/S3590/cfamily/rule.adoc (hunk 5 above)
Comment (LoghinVladDev) [reply rules/S3590/cfamily/rule.adoc:20]: I was thinking this is a bit of a stretch, but I have seen this happen, I'm sure it makes sense to us but I will add an example to be thorough

File: rules/S3590/cfamily/rule.adoc (hunk 16 above)
Comment (LoghinVladDev) [reply rules/S3590/cfamily/rule.adoc:123]: while it is a stack variable, the string data is supposed to be located in `.rodata` in most implementations, I believe.
I will check on godbolt

File: rules/S3590/cfamily/rule.adoc (hunk 2 above)
Comment (LoghinVladDev) [reply rules/S3590/cfamily/rule.adoc:14]: Was oscillating with this one:
* While yes, I believe that UB impact >> crash impact, I did not find any documented case of UB that does not act as a crash in the case of non-owned memory free, so most of the times the user will encounter a crash.
* While not stated in the rule, at least a sentence in the "...potential impact" specifying crashes is a good thing to add.

TL;DR will add a short section -> almost all UB in this case is a crash.

File: rules/S3590/cfamily/rule.adoc (hunk 16 above)
Comment (pdschbrt) [reply rules/S3590/cfamily/rule.adoc]: You're right, of course. The actual string is likely part of the program's `.rodata`.

File: rules/S3590/cfamily/rule.adoc (hunk 5 above)
Comment (LoghinVladDev) [reply rules/S3590/cfamily/rule.adoc:20]: Edit: I cannot decide whether this is a 'pitfall' or a 'going the extra mile'. Will add it to 'going...' for now, tell me after the re-review if it fits better in 'pitfalls'

File: rules/S3590/cfamily/rule.adoc (hunk 8 above)
Comment (LoghinVladDev) [reply rules/S3590/cfamily/rule.adoc:46]: I would use something a bit more generic, since the example also contains `alloca`

File: rules/S3590/cfamily/rule.adoc (hunk 11 above)
Comment (LoghinVladDev) [reply rules/S3590/cfamily/rule.adoc:74]: Added

File: rules/S3590/cfamily/rule.adoc (hunk 14 above)
Comment (LoghinVladDev) [reply rules/S3590/cfamily/rule.adoc:110]: Added

File: rules/S3590/cfamily/rule.adoc (hunk 3 above)
Comment (LoghinVladDev) [reply rules/S3590/cfamily/rule.adoc]: Reprashed

**Comment by LoghinVladDev** [comment]
Added a 'small' "Going the extra mile" section.
Do tell if it should be reduced.

File: rules/S3590/cfamily/rule.adoc [hunk 17]
```diff
@@ -1,38 +1,245 @@
...
+  char const* name = "string";
+  // ...
+  // Compliant: read-only program data is freed at the end of the program's execution.
+}
+----
+
+
+=== Going the extra mile
+
+Freeing stack memory usually happens if the origin of the memory held by a pointer is mixed.
```
Comment (LoghinVladDev) [inline rules/S3590/cfamily/rule.adoc:149]: Used "origin of the memory held by a pointer is mixed." since I want to avoid "origin of the memory pointed at by a pointer". Please indicate if you see a better way to phrase this

File: rules/S3590/cfamily/rule.adoc (hunk 2 above)
Comment (pdschbrt) [reply rules/S3590/cfamily/rule.adoc:14]: Please also align with the definition of UB that we now use in many rules. You can find an example in the description of rule `S3588`, for instance.
```
When a program comprises undefined behavior, the compiler no longer needs to adhere to the language standard, and the program has no meaning assigned to it.
The application might just crash, but in the worst case, the application may appear to execute correctly, while losing data or producing incorrect results.
```
This should be part of the `What is the potential impact` section.

File: rules/S3590/cfamily/rule.adoc (hunk 8 above)
Comment (pdschbrt) [reply rules/S3590/cfamily/rule.adoc:49]: Sounds good to me, too.

File: rules/S3590/cfamily/rule.adoc [hunk 18]
```diff
@@ -1,38 +1,245 @@
+Explicitly releasing non-heap memory leads to undefined behavior.
+
 == Why is this an issue?
 
-Stack allocated memory, like memory allocated with the functions ``++alloca++``, ``++_alloca++``, ``++_malloca++``, ``++__builtin_alloca++``, is automatically released at the end of the function, and should not be released with ``++free++``. Explicitly ``++free++``-ing such memory results in undefined behavior.
+The `free` function and `delete` operator are used exclusively to release dynamically allocated memory.
+Attempting to release any other type of memory is _undefined behavior_.
+
+The following non-heap memory types should not be released:
```
Comment (pdschbrt) [inline rules/S3590/cfamily/rule.adoc:8]:
```suggestion
The following non-heap memory types may not be released:
```
Makes it a bit stronger than `should` in this context.

File: rules/S3590/cfamily/rule.adoc [hunk 19]
```diff
@@ -1,38 +1,245 @@
...
+  char const* name = "string";
+  // ...
+  // Compliant: read-only program data is freed at the end of the program's execution.
+}
+----
+
+
+=== Going the extra mile
+
+Freeing stack memory usually happens if the origin of the memory held by a pointer is mixed.
```
Comment (pdschbrt) [inline rules/S3590/cfamily/rule.adoc:149]: How about simplifying this a bit like in:
```suggestion
The accidental release of non-heap memory usually occurs in practice if the same pointer variable is used to once reference heap and once non-heap memory.
This may lead to confusion and should be avoided.
```

File: rules/S3590/cfamily/rule.adoc [hunk 20]
```diff
@@ -1,38 +1,245 @@
...
+Specifically if the same pointer is used to reference heap and non-heap memory.
+
+This can happen
```
Comment (pdschbrt) [inline rules/S3590/cfamily/rule.adoc:152]:
```suggestion

```

File: rules/S3590/cfamily/rule.adoc [hunk 21]
```diff
@@ -1,38 +1,245 @@
...
+
+To avoid this, follow these best practices:
```
Comment (pdschbrt) [inline rules/S3590/cfamily/rule.adoc:154]:
```suggestion
These best practices help to avoid accidentally releasing non-heap memory:
```

File: rules/S3590/cfamily/rule.adoc [hunk 22]
```diff
@@ -1,38 +1,245 @@
...
+* If accessing different memory types, different pointers should be used.
```
Comment (pdschbrt) [inline rules/S3590/cfamily/rule.adoc:155]:
```suggestion
* If accessing different memory types, use different pointer variables.
```

File: rules/S3590/cfamily/rule.adoc [hunk 23]
```diff
@@ -1,38 +1,245 @@
...
+* When passing non-heap memory address to a function, ensure that the function does not attempt to release the memory.
```
Comment (pdschbrt) [inline rules/S3590/cfamily/rule.adoc:156]:
```suggestion
* When passing non-heap memory addresses to functions, ensure that the functions do not attempt to release the memory.
```

File: rules/S3590/cfamily/rule.adoc [hunk 24]
```diff
@@ -1,38 +1,245 @@
...
+void fun(int length) {
+  static char smallString[32];
+  char* usedString;
+
+  if (length < 31) {
+    usedString = smallString; // Pointer to stack memory assigned here
+  } else {
+    usedString = (char*)malloc(length + 1);
+  }
+  useAndClear(usedString); // If length < 31, the unsafe memory will free memory located on the stack.
```
Comment (pdschbrt) [inline rules/S3590/cfamily/rule.adoc:219]: `useAndClear` is not defined here.

File: rules/S3590/cfamily/rule.adoc [hunk 25]
```diff
@@ -1,38 +1,245 @@
...
+This can happen
+
+To avoid this, follow these best practices:
+* If accessing different memory types, different pointers should be used.
+* When passing non-heap memory address to a function, ensure that the function does not attempt to release the memory.
+* If manually managing dynamic memory, release it in the same scope where it was acquired.
+
+The following example illustrates the usage of different pointers:
+
+Noncompliant code example
```
Comment (pdschbrt) [inline rules/S3590/cfamily/rule.adoc:161]: I would either make these subsections of this very section or use short sentences to briefly explain what the respective code snippet shows. For instance,
```
The following piece of code shows a situation in which the same pointer variable is used to hold a stack or heap address, which leads to a situation in which heap memory is accidentally released.
```
I would prefer the later option, but leave the decision up to you.

**Review by pdschbrt — CHANGES_REQUESTED** [review]
Some additional remarks.

File: rules/S3590/cfamily/rule.adoc (hunk 18 above)
Comment (LoghinVladDev) [reply rules/S3590/cfamily/rule.adoc:8]: I believe `must` is the intention, and I will go with that. The following ... `may` not ....; for me, implies the possibility that it might not happen

File: rules/S3590/cfamily/rule.adoc (hunk 20 above)
Comment (LoghinVladDev) [reply rules/S3590/cfamily/rule.adoc:152]: Whoops

File: rules/S3590/cfamily/rule.adoc (hunk 2 above)
Comment (LoghinVladDev) [reply rules/S3590/cfamily/rule.adoc:14]: Then should we start putting this into a new `shared_content/cfamily/potential_impact_undefined_behavior.adoc`?

File: rules/S3590/cfamily/rule.adoc (hunk 18 above)
Comment (pdschbrt) [reply rules/S3590/cfamily/rule.adoc:8]: In this context "may" means "disallowed". But I'm fine with "must", too.

File: rules/S3590/cfamily/rule.adoc (hunk 2 above)
Comment (pdschbrt) [reply rules/S3590/cfamily/rule.adoc:14]: Ideally, yes. However, many rules have slightly adjusted this baseline explanation of UB. So in practice, we cannot really share this.

File: rules/S3590/cfamily/rule.adoc (hunk 18 above)
Comment (LoghinVladDev) [reply rules/S3590/cfamily/rule.adoc:8]: Actually yes, you are correct, re-read and agree

File: rules/S3590/cfamily/rule.adoc (hunk 25 above)
Comment (LoghinVladDev) [reply rules/S3590/cfamily/rule.adoc:161]: Strongly agree with describing more in detail, tracks better to comments (and edit those a bit)

File: rules/S3590/cfamily/rule.adoc (hunk 24 above)
Comment (LoghinVladDev) [reply rules/S3590/cfamily/rule.adoc]: addressed

File: rules/S3590/cfamily/rule.adoc [hunk 26]
```diff
@@ -1,38 +1,248 @@
...
+
+The following non-heap memory types may not be released:
+
+* Stack allocated memory - local variables or memory allocated with the `alloca`, `_alloca`, `_malloca` and `__builtin_alloca` functions.
+* Executable program code - function pointers.
+* Program data - global and static variables.
+* Read-only program data - constants and strings.
+
+=== What is the potential impact
+
```
[the remark this answers is not in the document]
Comment (LoghinVladDev) [reply rules/S3590/cfamily/rule.adoc:16]: I truly believe github should include some "10 lines above/10 lines below" for the suggestions as well, I stared a bit at the suggestion :laughing:. 

Regardless, is this not repeating the same stated concept at line 6 - "Attempting to release any other type of memory is _undefined behavior_?

File: rules/S3590/cfamily/rule.adoc [hunk 27]
```diff
@@ -1,38 +1,248 @@
...
+[source,cpp,diff-type=compliant,diff-id=2]
+----
+int get10() {
+  return 10;
+}
+
+void fun() {
   // ...
-  free(name2); // Noncompliant, memory allocated on the stack
+  // Compliant: program code will be released at the end of the program's execution.
```
[the remark this answers is not in the document]
Comment (LoghinVladDev) [reply rules/S3590/cfamily/rule.adoc:80]: Didn't think of that, that's a great suggestion

File: rules/S3590/cfamily/rule.adoc (hunk 26 above)
[the remark this answers is not in the document]
Comment (pdschbrt) [reply rules/S3590/cfamily/rule.adoc:16]: That's fine. For most, if not all, of the other rules, we also give a very quick repetition/summary of what goes wrong "IF ..." as an introduction, and then elaborate on what that means for the program and explain the consequences.

[the remark this answers is not in the document]
Comment (LoghinVladDev) [reply rules/S3590/cfamily/rule.adoc:16]: I truly want to be able to have in-rule links, and have line 6 UB be a link to "potential impact" UB, so we don't have to second guess clarity, since I'm really 50-50 repeats like these, and I balanced the scales with your opinion

**Comment by LoghinVladDev** [comment]
Apologies for the occasional "leftovers", and thanks for observing them! Was fighting with SQ and getting an informer to display for the longest time today. I'll give it another read since it's past the demo

File: rules/S3590/cfamily/rule.adoc (hunk 25 above)
Comment (pdschbrt) [reply rules/S3590/cfamily/rule.adoc:169]: Did you already address this, @LoghinVladDev?
If I'm not mistaken, you elaborated on the code comments. In that case, use the appropriate number of `=` to make the `Noncompliant code example` and `Compliant solution` (sub)section headers.

File: rules/S3590/cfamily/rule.adoc [hunk 28]
```diff
@@ -1,38 +1,254 @@
...
+  static char smallString[32];
+  if (length < 31) {
+    use(smallString);
+  } else {
+    heapString = (char*)malloc(length + 1);
+    use(heapString);
+    free(heapString); // Compliant: memory released in the scope it was acquired in.
+  }
+}
+----
```
Comment (pdschbrt) [inline rules/S3590/cfamily/rule.adoc:252]: Do we have some materials concerning undefined behavior, for instance, that we could link here?

**Review by pdschbrt — CHANGES_REQUESTED** [review]
Almost done.

File: rules/S3590/cfamily/rule.adoc (hunk 25 above)
Comment (LoghinVladDev) [reply rules/S3590/cfamily/rule.adoc:169]: Oh, I completely missed this, as in, the `==== Compliant ....` formatting you were referring to.

I did not add the `====` since Noncompliant/compliant appears only in How to fix it as subheaders, not in the Going the extra mile section.

I will check if `====` is allowed (if CI formatting check passes)

File: rules/S3590/cfamily/rule.adoc (hunk 28 above)
Comment (LoghinVladDev) [reply rules/S3590/cfamily/rule.adoc:252]: No materials of UB caused by stack free found, only crashes.

File: rules/S3590/cfamily/rule.adoc (hunk 25 above)
Comment (LoghinVladDev) [reply rules/S3590/cfamily/rule.adoc]: Update: yep, CI tooling fails with ==== inside Going the Extra Mile

File: rules/S3590/cfamily/rule.adoc (hunk 28 above)
Comment (pdschbrt) [reply rules/S3590/cfamily/rule.adoc:252]: Ok. Leaving this blank should be fine, then.

**Review by pdschbrt — APPROVED** [review]
Looks good now.

**Comment by LoghinVladDev** [comment]
Merged together with https://github.com/SonarSource/sonar-cpp/pull/2991
